### PR TITLE
Bucket and files

### PR DESCRIPTION
**Pull request by mosch**
Added bucket and files to the SDK. 

To make it work I need to do two changes to the request: 
* Allowing delete method to have a body
* Allowing to pass headers and form-data body

### CODE REVIEW

File: src/rest/methods/file.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  file: Blob | Buffer | ReadableStream,
+  name: string,
+  path?: string,
+) => FileResult
+export async function fileCreate(
+  client: InterfaceAllthingsRestClient,
+  file: Blob | Buffer | ReadableStream,
+  name: string,
+  path?: string,
+): FileResult {
```
[the remark this answers is not in the document]
Comment (mosch) [reply src/rest/methods/file.ts]: What's the point of this? It requires the developer to write the object-key every time.

**Comment by mosch** [comment]
> I have some concern about the naming of bucketDeleteFile(), specifically I'm hung up on the word "delete" because it sounds like I've destroyed the file, but actually I've only removed it from the list of files in that bucket. This may be misleading. 🤔

I agree. Let's tune the rules for naming:

**For Relations**
Add / Remove
_or_
Link / Unlink

**For C&D**
Create / Delete

I'd prefer add / remove but it think it could be also confusing to use delete and remove.

File: src/rest/methods/bucket.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+): BucketResult {
+  return client.post(`/v1/buckets/${bucketId}/files`, {
+    id: fileId,
+  })
+}
+
+export type MethodBucketDeleteFile = (
+  bucketId: string,
+  fileId: string,
+) => Promise<string>
```
[the remark this answers is not in the document]
Comment (iamn1ck) [reply src/rest/methods/bucket.ts]: Agreed, refer to this function for deleting registration codes
https://github.com/allthings/node-sdk/blob/bucket-and-files/src/rest/methods/registrationCode.ts#L86

File: src/rest/methods/bucket.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  return client.post('/v1/buckets', {
+    channels: data.channels,
+    name: data.name,
+  })
+}
+
+export type MethodBucketAddFile = (
+  bucketId: string,
+  fileId: string,
+) => Promise<string>
```
Comment (adieuadieu) [inline src/rest/methods/bucket.ts]:
```suggestion
) => Promise<boolean>
```

File: src/rest/methods/bucket.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+export async function bucketAddFile(
+  client: InterfaceAllthingsRestClient,
+  bucketId: string,
+  fileId: string,
+): Promise<string> {
```
Comment (adieuadieu) [inline src/rest/methods/bucket.ts]:
```suggestion
): Promise<boolean> {
```

File: src/rest/methods/bucket.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  return client.post(`/v1/buckets/${bucketId}/files`, {
```
Comment (adieuadieu) [inline src/rest/methods/bucket.ts]:
```suggestion
  return (await client.post(`/v1/buckets/${bucketId}/files`, {
```

File: src/rest/methods/bucket.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    id: fileId,
+  })
```
Comment (adieuadieu) [inline src/rest/methods/bucket.ts]:
```suggestion
  })) === ''
```

File: src/rest/methods/file.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  const data = new FormData()
+  // tslint:disable-next-line:no-expression-statement
+  data.append('file', file, name)
+  // tslint:disable-next-line:no-expression-statement
+  data.append('path', path || '')
+
+  return client.post('/v1/files', data, data.getHeaders())
+}
+
+export type MethodFileDelete = (fileId: string) => string
```
[the remark this answers is not in the document]
Comment (adieuadieu) [reply src/rest/methods/file.ts]: Looks like this function signature is still incorrect. `Promise<boolean>` probably makes most sense on success.

It's also incorrect down on line 59.

File: src/rest/request.ts [hunk 8]
```diff
@@ -132,19 +135,41 @@ export function makeApiRequest(
               : ''
           }`
 
+          const body = payload && payload.body
+          const hasForm = !!(body && body.formData)
+          const form = body && hasForm ? body.formData : {}
+          const formData = Object.keys(form).reduce((prev, name) => {
```
Comment (adieuadieu) [inline src/rest/request.ts]: `previous` doesn't really need to be abbreviated.

You might alternatively consider using `Object.entries()` here instead. Then you don't need to access the value by key on `form`. You'll already have it.

File: src/rest/request.ts [hunk 9]
```diff
@@ -132,19 +135,41 @@ export function makeApiRequest(
...
+          }, new FormDataModule())
+
+          const requestHeaders = {
+            ...(hasForm && typeof formData.getHeaders === 'function'
+              ? formData.getHeaders()
+              : {}),
+          }
+
+          const requestBody = {
+            // Sorry typescript you got it wrong this time
```
Comment (adieuadieu) [inline src/rest/request.ts]: This comment isn't so helpful. Nor is it exactly correct. The reality is more complex.

File: src/rest/types.ts [hunk 10]
```diff
@@ -190,6 +221,18 @@ export interface InterfaceAllthingsRestClient {
...
 
+  // File
+
+  /**
+   * Creates a file
+   */
+  readonly fileCreate: MethodFileCreate
+
+  /**
+   * Delets a file
```
Comment (adieuadieu) [inline src/rest/types.ts]:
```suggestion
   * Deletes a file by it's ID
```

**Pull request merged**